### ISSUE TITLE
Installation error on R by `Could NOT find OpenMP_C`

### ISSUE THREAD

**Issue by JiaxiangBU** [microsoft/LightGBM#2026]
I search answers from Stack Overflow (https://stackoverflow.com/questions/tagged/lightgbm) but unfortunately get the solution. Here are related solutions.

* [cannot install lightgbm in R with devtools on macOS - Stack Overflow](https://stackoverflow.com/questions/50704874/cannot-install-lightgbm-in-r-with-devtools-on-macos?r=SearchResults)
* [cannot install lightgbm in R with devtools on macOS - Stack Overflow](https://stackoverflow.com/questions/50704874/cannot-install-lightgbm-in-r-with-devtools-on-macos/51005728?r=SearchResults#51005728)

## Environment info

Operating System: macOS Mojave (10.14.1)

CPU/GPU model: CPU

C++/Python/R version: R

LightGBM version or commit hash:

## Error message

```
$ Rscript build_r.R
   checking for file ‘/Users/vija/Downloads/work/LightGBM/lightgbm_r/DESCRIPTION✔  checking for file ‘/Users/vija/Downloads/work/LightGBM/lightgbm_r/DESCRIPTION’
─  preparing ‘lightgbm’:
✔  checking DESCRIPTION meta-information ...
─  cleaning src
─  checking for LF line-endings in source and make files and shell scripts
─  checking for empty or unneeded directories
   WARNING: directory ‘lightgbm/src/compute’ is empty
─  looking to see if a ‘data/datalist’ file should be added
─  building ‘lightgbm_2.2.4.tar.gz’
   
[1] "/Users/vija/Downloads/work/LightGBM/lightgbm_2.2.4.tar.gz"
* installing to library ‘/Library/Frameworks/R.framework/Versions/3.5/Resources/library’
* installing *source* package ‘lightgbm’ ...
** libs
clang -I"/Library/Frameworks/R.framework/Resources/include" -DNDEBUG   -I/usr/local/include   -fPIC  -Wall -g -O2  -c R_init.c -o R_init.o
clang -dynamiclib -Wl,-headerpad_max_install_names -undefined dynamic_lookup -single_module -multiply_defined suppress -L/Library/Frameworks/R.framework/Resources/lib -L/usr/local/lib -o lightgbm.so R_init.o -F/Library/Frameworks/R.framework/.. -framework R -Wl,-framework -Wl,CoreFoundation
installing via 'install.libs.R' to /Library/Frameworks/R.framework/Versions/3.5/Resources/library/lightgbm
-- The C compiler identification is AppleClang 10.0.0.10001044
-- The CXX compiler identification is AppleClang 10.0.0.10001044
-- Check for working C compiler: /Library/Developer/CommandLineTools/usr/bin/cc
-- Check for working C compiler: /Library/Developer/CommandLineTools/usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /Library/Developer/CommandLineTools/usr/bin/c++
-- Check for working CXX compiler: /Library/Developer/CommandLineTools/usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Error at /usr/local/Cellar/cmake/3.12.3/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:137 (message):
  Could NOT find OpenMP_C (missing: OpenMP_C_FLAGS OpenMP_C_LIB_NAMES)
Call Stack (most recent call first):
  /usr/local/Cellar/cmake/3.12.3/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:378 (_FPHSA_FAILURE_MESSAGE)
  /usr/local/Cellar/cmake/3.12.3/share/cmake/Modules/FindOpenMP.cmake:466 (find_package_handle_standard_args)
  CMakeLists.txt:81 (find_package)


-- Configuring incomplete, errors occurred!
See also "/private/var/folders/yt/307v0c_945z1rh043jsdx1lh0000gn/T/RtmpSrIuyS/R.INSTALL88ff3951e6c4/lightgbm/src/build/CMakeFiles/CMakeOutput.log".
make: *** No rule to make target `_lightgbm'.  Stop.
Error in eval(ei, envir) : Cannot find lib_lightgbm.so
* removing ‘/Library/Frameworks/R.framework/Versions/3.5/Resources/library/lightgbm’
```

## Reproducible examples

```
brew install cmake
brew install libomp
git clone --recursive https://github.com/Microsoft/LightGBM ; cd LightGBM
mkdir build ; cd build

# For Mojave (10.14)
cmake \
  -DOpenMP_C_FLAGS="-Xpreprocessor -fopenmp -I$(brew --prefix libomp)/include" \
  -DOpenMP_C_LIB_NAMES="omp" \
  -DOpenMP_CXX_FLAGS="-Xpreprocessor -fopenmp -I$(brew --prefix libomp)/include" \
  -DOpenMP_CXX_LIB_NAMES="omp" \
  -DOpenMP_omp_LIBRARY=$(brew --prefix libomp)/lib/libomp.dylib \
  ..
cd ../
Rscript build_r.R
```

## Steps to reproduce

1.  Run all code above. Reference from [LightGBM/R-package at master · Microsoft/LightGBM](https://github.com/Microsoft/LightGBM/tree/master/R-package)
2. The key error is `Could NOT find OpenMP_C (missing: OpenMP_C_FLAGS OpenMP_C_LIB_NAMES)`. I think the OpenMP path is unknown for LightGBM here.

**Comment by StrikerRUS** [comment]
> If you are using a precompiled dll/lib locally, you can move the dll/lib into LightGBM root folder, modify LightGBM/R-package/src/install.libs.R's 2nd line (change use_precompile <- FALSE to use_precompile <- TRUE), and install R-package as usual.

**Comment by jameslamb** [comment]
@JiaxiangBU thanks for reporting this. I noticed you're compiling with `clang` instead of `g++` and that you've followed our [installation notes for Mac](https://github.com/Microsoft/LightGBM/tree/master/R-package#mac-os-preparation). So I'm not sure what's happening, but we'll figure it out!

@StrikerRUS I feel like this is the same issue as #1898 . I'm confused reading through that issue...it seems like the OP was able to resolve the issue on Mojave by manually changing `CmakeLists.txt`, is that correct? And then #1923 should have fixed this?

**Comment by jameslamb** [comment]
or @StrikerRUS are you recommending that @JiaxiangBU take `liblightgbm.so` (or whatever it ends up being called) and just take the "run from a precompiled library" path?

**Comment by Laurae2** [comment]
@jameslamb the OP is trying to explicitly precompile LightGBM before installing LightGBM R package. Except when precompiling I think it should not be done using build_r.R but directly the standard install package.

**Comment by StrikerRUS** [comment]
@jameslamb @Laurae2 I think that we definitely need a note in README about that installation on Mojave is possible only via precompile method now.

Or can we add the same logic in https://github.com/Microsoft/LightGBM/blob/master/R-package/src/install.libs.R as we added in #1923 and #1975 for Python-package?

**Comment by Laurae2** [comment]
ping @jameslamb for opinion (I don't use Mojave so I'm from a blind view perspective)

**Comment by StrikerRUS** [comment]
@Laurae2 @jameslamb I've [proposed changes](https://github.com/Microsoft/LightGBM/issues/2086) in our CI for the compatibility with Mojave: everything works well. Python installation (script and docs) and CLI-version installation guide have been already patched. So what about R?

**Comment by jameslamb** [comment]
Will come back to this tonight, sorry for the delay

**Comment by Laurae2** [comment]
ping @jameslamb

**Comment by jameslamb** [comment]
I believe #2164 will address this for the R package, @StrikerRUS .

**Comment by StrikerRUS** [comment]
@jameslamb Yeah, I hope! 😃

**Comment by StrikerRUS** [comment]
Fixed in #2164.
